pyteal/ast/router_test.py

```diff
@@ -0,0 +1,533 @@
+
+        args: list[pt.abi.BaseType] = [
+            spec.new_instance()
+            for spec in typing.cast(
+                list[pt.abi.TypeSpec], abi_subroutine.subroutine.expected_arg_types
+            )
+        ]
+
+        loading: list[pt.Expr]
+
+        if abi_subroutine.subroutine.argument_count() > pt.METHOD_ARG_NUM_LIMIT:
+            sdk_last_arg = pt.abi.TupleTypeSpec(
+                *[
+                    spec
+                    for spec in typing.cast(
+                        list[pt.abi.TypeSpec],
+                        abi_subroutine.subroutine.expected_arg_types,
+                    )[pt.METHOD_ARG_NUM_LIMIT - 1 :]
+                ]
+            ).new_instance()
+            loading = [
+                arg.decode(pt.Txn.application_args[index + 1])
+                for index, arg in enumerate(args[: pt.METHOD_ARG_NUM_LIMIT - 1])
+            ]
+            loading.append(
+                sdk_last_arg.decode(pt.Txn.application_args[pt.METHOD_ARG_NUM_LIMIT])
+            )
+            for i in range(pt.METHOD_ARG_NUM_LIMIT - 1, len(args)):
+                loading.append(
+                    sdk_last_arg[i - pt.METHOD_ARG_NUM_LIMIT + 1].store_into(args[i])
+                )
+        else:
+            loading = [
+                arg.decode(pt.Txn.application_args[index + 1])
+                for index, arg in enumerate(args)
+            ]
+
+        evaluate: pt.Expr
+        if abi_subroutine.type_of() != "void":
+            output_temp = abi_subroutine.output_kwarg_info.abi_type.new_instance()
+            evaluate = pt.Seq(
+                abi_subroutine(*args).store_into(output_temp),
+                pt.abi.MethodReturn(output_temp),
+            )
+        else:
+            evaluate = abi_subroutine(*args)
+
+        actual = assemble_helper(pt.Seq(*loading, evaluate, pt.Approve()))
+        with pt.TealComponent.Context.ignoreScratchSlotEquality(), pt.TealComponent.Context.ignoreExprEquality():
+            assert actual == assembled_wrapped
+
+
+def test_contract_json_obj():
+    abi_subroutines = list(
+        filter(lambda x: isinstance(x, pt.ABIReturnSubroutine), GOOD_SUBROUTINE_CASES)
+    )
+    contract_name = "contract_name"
+    on_complete_actions = pt.BareCallActions(
+        clear_state=pt.OnCompleteAction.call_only(safe_clear_state_delete)
+    )
+    router = pt.Router(contract_name, on_complete_actions)
+    method_list: list[sdk_abi.Method] = []
+    for subroutine in abi_subroutines:
+        router.add_method_handler(subroutine)
+        method_list.append(sdk_abi.Method.from_signature(subroutine.method_signature()))
+    sdk_contract = sdk_abi.Contract(contract_name, method_list)
+    contract = router.contract_construct()
+    assert sdk_contract.desc == contract.desc
+    assert sdk_contract.name == contract.name
+    assert sdk_contract.networks == contract.networks
+    for method in sdk_contract.methods:
+        assert method in contract.methods
```

Comment:
Can you do `assert sdk_contract.methods == contract.methods` instead? That gives a stronger test of equality; the current code will not raise an alarm if `sdk_contract.methods` is empty when it shouldn't be.

Alternatively, calling `dictify()` on both contracts and comparing the result is also a possibility.

Reply:
that makes sense. The previous implementation was aligned with the thinking of either sorting AST, or concatenating AST components in specific order. The current one does not require ordering being changed, so we can directly apply equality check.